### Pair Programming ANOVA

En el pair programming de hoy usaremos el set de datos que guardastéis en el pair programming de normalización y estandarización.

Hasta ahora habéis estado evaluando las características de vuestro set de datos y habéis hecho una gran exploración, es el momento de hacer vuestro primer ANOVA! En el ejercicio de hoy tendréis que hacer un ANOVA con vuestro datos y hacer una interpretación de los resultados.

📌 NOTA Puede que vuestros datos no se ajusten o no cumplan todas las asunciones, no pasa nada, haced el ANOVA e interpretad los resultados. En próximas lecciones aprenderemos que podemos hacer cuando nos encontramos en esta situación.

In [1]:
# Tratamiento de datos
# -----------------------------------------------------------------------
import pandas as pd
import numpy as np
from sklearn.preprocessing import StandardScaler
import statsmodels.api as sm
from statsmodels.formula.api import ols
from statsmodels.multivariate.manova import MANOVA
from sklearn.preprocessing import StandardScaler
# Gráficos
# -----------------------------------------------------------------------
import seaborn as sns
import matplotlib.pyplot as plt

# Para la gestión de los warning
# ------------------------------------------------------------------------------
import warnings
warnings.filterwarnings('ignore')

# Para poder ver todas las columnas del dataframe
# ------------------------------------------------------------------------------
pd.options.display.max_columns = None 

In [4]:
#abrimos el df
df=pd.read_csv('datos/diamonds_est.csv', index_col=0)
df.head(3)

carat      cut color clarity  depth  table  price  lenght_mm  width_mm  \
1   0.23    Ideal     E     SI2   61.5   55.0    326       3.95      3.98   
2   0.21  Premium     E     SI1   59.8   61.0    326       3.89      3.84   
3   0.23     Good     E     VS1   61.8   65.0    327       4.05      4.07   

   depth_mm  est_carat  est_depth  est_table  est_price  est_lenght_mm  \
1      2.43  -1.268160  -1.597233   1.661056  -0.986397      -1.646799   
2      2.31  -1.224362   0.042616   3.519383  -0.986075      -1.503737   
3      2.31  -1.092965   0.534571   0.267310  -0.983820      -1.369617   

   est_width_mm  est_depth_mm  
1     -1.705394     -1.780396  
2     -1.498268     -1.780396  
3     -1.354180     -1.316852

In [6]:
#iniciamos nuestro anova:
lm = ols('carat ~ cut + color + clarity + depth + table + price + lenght_mm + width_mm + depth_mm', data=df).fit()
sm.stats.anova_lm(lm)

df       sum_sq      mean_sq              F         PR(>F)
cut            4.0   372.268622    93.067156    6216.078945   0.000000e+00
color          6.0   893.987192   148.997865    9951.765339   0.000000e+00
clarity        7.0  1192.812489   170.401784   11381.361512   0.000000e+00
depth          1.0     2.684554     2.684554     179.304931   7.992439e-41
table          1.0    81.791052    81.791052    5462.932990   0.000000e+00
price          1.0  5340.661544  5340.661544  356709.878542   0.000000e+00
lenght_mm      1.0  2511.694877  2511.694877  167759.437106   0.000000e+00
width_mm       1.0     2.605623     2.605623     174.032992   1.122290e-39
depth_mm       1.0    13.286903    13.286903     887.449918  1.979802e-193
Residual   53770.0   805.044627     0.014972            NaN            NaN

In [7]:
#extraemos un resumen de la recta que se ajusta a nuestro modelo:
lm.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  carat   R-squared:                       0.928
Model:                            OLS   Adj. R-squared:                  0.928
Method:                 Least Squares   F-statistic:                 3.024e+04
Date:                Mon, 23 Jan 2023   Prob (F-statistic):               0.00
Time:                        09:48:12   Log-Likelihood:                 36691.
No. Observations:               53794   AIC:                        -7.333e+04
Df Residuals:                   53770   BIC:                        -7.312e+04
Df Model:                          23                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept           -1.6287      0.051    -31.918      0.000      -1.729      -1.529
cut[T.Good]          0.0052      0.004      1.391      0.164      -0.002       0.012
cut[T.Ideal]         0.0058      0.004      1.647      0.100      -0.001       0.013
cut[T.Premium]       0.0123      0.003      3.592      0.000       0.006       0.019
cut[T.Very Good]     0.0096      0.003      2.740      0.006       0.003       0.016
color[T.E]           0.0030      0.002      1.562      0.118      -0.001       0.007
color[T.F]        4.025e-05      0.002      0.021      0.984      -0.004       0.004
color[T.G]           0.0070      0.002      3.616      0.000       0.003       0.011
color[T.H]           0.0200      0.002      9.784      0.000       0.016       0.024
color[T.I]           0.0351      0.002     15.229      0.000       0.031       0.040
color[T.J]           0.0240      0.003      8.461      0.000       0.018       0.030
clarity[T.IF]        0.0485      0.006      8.668      0.000       0.038       0.059
clarity[T.SI1]       0.0298      0.005      6.266      0.000       0.020       0.039
clarity[T.SI2]       0.0379      0.005      7.970      0.000       0.029       0.047
clarity[T.VS1]       0.0305      0.005      6.263      0.000       0.021       0.040
clarity[T.VS2]       0.0343      0.005      7.159      0.000       0.025       0.044
clarity[T.VVS1]      0.0477      0.005      9.215      0.000       0.038       0.058
clarity[T.VVS2]      0.0397      0.005      7.879      0.000       0.030       0.050
depth                0.0009      0.001      1.224      0.221      -0.001       0.002
table                0.0024      0.000      7.564      0.000       0.002       0.003
price             8.106e-06   2.93e-07     27.660      0.000    7.53e-06    8.68e-06
lenght_mm            0.1677      0.010     17.236      0.000       0.149       0.187
width_mm             0.0455      0.010      4.677      0.000       0.026       0.065
depth_mm             0.2605      0.009     29.790      0.000       0.243       0.278
==============================================================================
Omnibus:                    47874.659   Durbin-Watson:                   1.439
Prob(Omnibus):                  0.000   Jarque-Bera (JB):          5754700.199
Skew:                          -3.801   Prob(JB):                         0.00
Kurtosis:                      53.096   Cond. No.                     4.47e+05
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.47e+05. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

#### Variable predictora 'cut'

Observamos que en los valores extremos de la categoría **'cut'** (el menor... el mayor...) no existen diferencias significativas (p>0.05) entre el peor y mejor corte y el número de carats que tiene el diamante. Sin embargo, si existen en las categorías intermedias: 'Very good' (p=0.006) y 'Premium' (p=0.000).

Observamos que la relación entre las variables es directa (coef=0.0123), por lo que a mayor número de carats, mayor probabilidad de que el corte sea 'Premium'. La tasa de variabilidad de las medias de número de carats y corte 'Premium' es del 0.0123, por lo que por cada punto que varíe la media de carats, el corte premium variará un 0.01%. Es decir, existe relación pero ésta es ínfima.


#### Variable predictora 'color'

Observamos que en los valores extremos de la categoría **'color' (TE,TF)** no existen diferencias significativas (p>0.05) entre ambos colores y el número de carats que tiene el diamante. Sin embargo, si existen en en el resto de colores, p = 0.00. Es significativa en el resto de categorias medias-inferiores del color (G-J, inclusive).

Por lo tanto, salvo en el peor y mejor de los colores, cuanto más carats es el diamante, peor será la calidad de su color. Sin embargo, su coeficiente es muy bajo (menor de 0.5), por lo que esta variación es muy, muy ligera.


#### Variable predictora 'depth_mm' y length_mm

La relación que hay entre profundidad y longitud es directa y significativa (p<0.01). Por lo tanto, a mayor número de carats, más profundo y largo será el diamante.


**CONCLUSIÓN: la varianza en las medias de los numeros de 'carats' es multifactorial, dependiendo entre otros factores del corte, del color, pero especialmente de la profundidad (depth_mm) y del largo del diamante (lenght_mm)**.